cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -16,10 +16,192 @@
+        g = j;
+        break;
+      case 1:
+        F = (D & B) | ((~D) & C);
+        g = (5 * j + 1) % 16;
+        break;
+      case 2:
+        F = B ^ C ^ D;
+        g = (3 * j + 5) % 16;
+        break;
+      case 3:
+        F = C ^ (B | (~D));
+        g = (7 * j) % 16;
+        break;
+    }
+
+    F = F + A + hash_constants[j] + buffer_ints[g];
+    A = D;
+    D = C;
+    C = B;
+
+    uint32_t shift = shift_constants[((j / 16) * 4) + (j % 4)];
+    B              = B + ((F << shift) | (F >> (32 - shift)));
+  }
+
+  hash_state->hash_value[0] += A;
+  hash_state->hash_value[1] += B;
+  hash_state->hash_value[2] += C;
+  hash_state->hash_value[3] += D;
+
+  hash_state->buffer_length = 0;
+}
+
+template <typename Key>
+struct MD5Hash {
+  using argument_type = Key;
+
+  /**
+   * @brief Core MD5 element processing function
+   */
+  template <typename TKey>
+  void __device__ process(TKey const& key,
+                          uint32_t const len,
+                          md5_intermediate_data* hash_state,
+                          md5_hash_constants_type const* hash_constants,
+                          md5_shift_constants_type const* shift_constants) const
+  {
+    uint8_t const* data = reinterpret_cast<uint8_t const*>(&key);
+    hash_state->message_length += len;
+
+    if (hash_state->buffer_length + len < 64) {
+      thrust::copy_n(thrust::seq, data, len, hash_state->buffer + hash_state->buffer_length);
+      hash_state->buffer_length += len;
+    } else {
+      uint32_t copylen = 64 - hash_state->buffer_length;
+
+      thrust::copy_n(thrust::seq, data, copylen, hash_state->buffer + hash_state->buffer_length);
+      md5_hash_step(hash_state, hash_constants, shift_constants);
+
+      while (len > 64 + copylen) {
+        thrust::copy_n(thrust::seq, data + copylen, 64, hash_state->buffer);
+        md5_hash_step(hash_state, hash_constants, shift_constants);
+        copylen += 64;
+      }
+
+      thrust::copy_n(thrust::seq, data + copylen, len - copylen, hash_state->buffer);
+      hash_state->buffer_length = len - copylen;
+    }
+  }
+
+  template <typename T, typename std::enable_if_t<std::is_floating_point<T>::value>* = nullptr>
+  void __device__ operator()(T const& key,
+                             md5_intermediate_data* hash_state,
+                             md5_hash_constants_type const* hash_constants,
+                             md5_shift_constants_type const* shift_constants) const
+  {
+    if (isnan(key)) {
+      T nan = std::numeric_limits<T>::quiet_NaN();
+      process(nan, sizeof(T), hash_state, hash_constants, shift_constants);
+    } else if (key == T{0.0}) {
+      process(T{0.0}, sizeof(T), hash_state, hash_constants, shift_constants);
+    } else {
+      process(key, sizeof(T), hash_state, hash_constants, shift_constants);
+    }
+  }
+
+  template <typename T,
+            typename std::enable_if_t<std::is_same<T, cudf::string_view>::value>* = nullptr>
```

Comment:
Use explicit specialization instead of SFINAE for single types. https://github.com/rapidsai/cudf/blob/branch-0.15/cpp/docs/TRANSITIONGUIDE.md#type-dispatcher